analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs

```diff
@@ -27,12 +27,27 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn.OperationProcessors
 {
     internal static class Binary
     {
-        public static ProgramState Process(SymbolicContext context, IBinaryOperationWrapper binary) =>
-            BinaryConstraint(binary.OperatorKind, context.State[binary.LeftOperand], context.State[binary.RightOperand]) is { } newConstraint
-                ? context.SetOperationConstraint(newConstraint)
-                : context.State;
+        public static ProgramState[] Process(SymbolicContext context, IBinaryOperationWrapper binary)
+        {
+            if (BinaryConstraint(binary.OperatorKind, context.State[binary.LeftOperand], context.State[binary.RightOperand]) is { } newConstraint)
+            {
+                return new[] { context.SetOperationConstraint(newConstraint) };   // We already know the answer from existing constraints
+            }
+            else
+            {
+                var positive = LearnBranchingConstraint(context.State, binary, false);
+                var negative = LearnBranchingConstraint(context.State, binary, true);
+                return positive == context.State && negative == context.State
+                    ? new[] { context.State }   // We can't learn anything, just move on
+                    : new[]
+                    {
+                        positive.SetOperationConstraint(context.Operation, BoolConstraint.True),
+                        negative.SetOperationConstraint(context.Operation, BoolConstraint.False)
```

Comment:
comprehension question: How do we make sure, that the right states flow into the right branches of the cfg in this case
```cs
if (arg == null)
{
  // true branch with the state from "positive"
}
```

Reply:
For `arg == null`, we build `BoolConstraint` for the whole operation. So it reduces to a question, how we handle `bool arg`
```
if(arg)
{
  // only when arg==True
}
```
Inside `ProcessBranching` of a node, we follow `Successors`. For each, we check if it is reachable (best to our knowledge)
https://github.com/SonarSource/sonar-dotnet/blob/9683a14b27b83e761907510f5962d018938b7397/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs#L115

Where branch is reachable when:
https://github.com/SonarSource/sonar-dotnet/blob/9683a14b27b83e761907510f5962d018938b7397/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs#L323-L330
it's unconditional, we have no idea about the condition or we have idea about it's BoolConstraint.

The logic is bit complicated, because we need to handle those "WhenTrue", "WhenFalse" and "Else" branch semantics.

Reply:
Ah thanks. Now it is more clear to me. I have seen the "IsReachable" call but didn't followed it further. So I never *reached* the BoolConstraint handling.